Algorithm.md

```diff
@@ -0,0 +1,30 @@
+1. Fetch available versions for each dependency
+2. If a project is already in the graph, intersect the new requirement with the one already present
+3. Select the highest [precedence] version which satisfies the intersected requirement (or only requirement, if this is the first insertion)
+  * This always has to be O(n) over the list of versions, because custom requirements mean that a previously failing version might become a candidate again
+4. Collect all child dependencies from that graph depth, repeat from step 1
+
+### In event of a conflict
+
+Given a project already in the graph with requirement X, and about to be re-added to the graph with requirement Y, where X and Y conflict:
+
+1. Find the _dependent_ project which changed the requirement into X (either from nil, or from a previously looser requirement)
+2. Select the next best version of that dependent project and attempt to finish resolving
```

Comment:
Hmm… maybe that will work for version ranges, but I'm not sure how generalizable it is.

For example, it could be that a particular _commit hash_ was rejected as a possible solution, and now we need to blacklist it. Although we could represent that with a requirement specifically blacklisting that one hash, is that really better than just iterating the list of available versions we have? I don't know the answer.